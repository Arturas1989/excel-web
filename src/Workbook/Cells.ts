import CSS from 'csstype';
import { type CellType, Row } from '../types';




export class Cells {
  public grid: Row[];
  private cellStyles: CSS.Properties;
  private contentStyles: CSS.Properties;
  private selectedStyles: CSS.Properties;
  private screenWidth: number;
  private screenHeight: number;
  private cellPadding: number;
  private borderWidth: number;
  private contentHeight: number;
  private contentPaddingleft: number;
  private cellWidth: number;
  private cellHeight: number;
  private rows: number;
  private cols: number;

  constructor() {
    this.cellPadding = 8;
    this.contentHeight = 16;
    this.cellWidth = 100;
    this.borderWidth = 1;
    this.contentPaddingleft = 1;
    this.selectedStyles = {};
    this.cellHeight = this.contentHeight + (this.cellPadding + this.borderWidth) * 2;
    this.cellStyles = {
      padding: this.cellPadding + 'px',
      width: this.cellWidth + 'px',
      border: `${this.borderWidth}px solid rgb(212, 212, 212)`,
      height: this.cellHeight + 'px',
      overflow: 'clip'
    };
    this.contentStyles = {
      fontSize: this.contentHeight + 'px',
      paddingLeft: this.contentPaddingleft + 'px',
    };

    this.screenWidth = window.innerWidth;
    this.screenHeight = window.innerHeight;
    this.rows = Math.ceil(this.screenHeight / this.cellHeight);
    this.cols = Math.ceil(this.screenWidth / this.cellWidth);

    this.grid = this.makeGrid();

    this.updateProperties = this.updateProperties.bind(this);
    window.addEventListener('resize', this.updateProperties);
  }

  updateProperties() {
    
    if(window.innerWidth > this.screenWidth){
      this.screenWidth = window.innerWidth;
      this.screenHeight = window.innerHeight;
      this.rows = Math.round(this.screenHeight / this.cellHeight);
      this.cols = Math.round(this.screenWidth / this.cellWidth);
      this.grid = this.makeGrid();
    }
  }

  

  makeGrid(): Row[]{
    console.log(this.cellHeight)
    let grid = [];
    for(let i = 0; i < this.rows; ++i){
      let row = [];
      for(let j = 0; j < this.cols; ++j){
        row.push({
          cellStyles: this.cellStyles,
          contentStyles: this.contentStyles,
          selected: '',
          address: this.getColName(j) + (i + 1),
          content: '',
          contentPaddingLeft: this.contentPaddingleft,
          row: i,
          col: j,
        } as CellType)
      }
      grid.push(row);
    }
    return grid;
  }

  getColName(col: number): string{
    const letters = 'ABCDEFGHIJKLMNOPQRSTUVWXYZ';
    let colLetters = [];
    while(col !== -1){
      const rem = col % 26;
      colLetters.push(letters[rem]) ;
      col = (col - rem) / 26 - 1;
    }
    return colLetters.reverse().join('');
  }
}
