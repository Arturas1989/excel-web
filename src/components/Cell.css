.cell-container{
  display: flex;
  flex-direction: column;
  width: 100dvw;
}

.cell-row{
  display: flex;
}

.cell > .content:focus-visible {
  outline: none;
}

.cell {
  width: var(--cell-width);
  height: calc(var(--cell-font-size) + 2 * (var(--cell-space) + var(--cell-selected-border-size)));
  border: 1px solid black;
  white-space: nowrap;
  user-select: none;
  padding: var(--cell-space);
  border: var(--cell-border);
  display: flex;
  flex-shrink: 0;
}

.cell > .content {
  width:fit-content;
  font-size: var(--cell-font-size);
  line-height: 1;
}

.cursorText {
  cursor: text;
}

.cursorCell{
  cursor: cell;
}

.selected{
  border: var(--cell-selected-border);
}
