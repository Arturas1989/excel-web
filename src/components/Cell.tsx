import { memo, useRef, useState } from 'react';
import { setCursorPositionInACell } from '../utils/helpers/cellHelper';
import { type DivInput, CellType, Methods } from '../types';
import CSS from 'csstype';

type CellProps = {
  cell: CellType;
  methods: Methods;
  cellStyle: CSS.Properties;
};

const InitialCell = ({ cell, methods, cellStyle }: CellProps) => {
  const cellRef = useRef<DivInput>(null);
  const [cursor, setCursor] = useState('cursorCell');
  const {
    contentStyles,
    address,
    content,
    contentPaddingLeft,
  } = cell;
  
  const {changeSelected} = methods;
  if(address === 'A1')console.log(cellStyle)

  function enableEditing(e: React.MouseEvent<HTMLDivElement, MouseEvent>) {
    setCursor('cursorText');
    selectCell();
    if (cellRef.current!.contentEditable === 'true') return;
    cellRef.current!.contentEditable = 'true';
    cellRef.current!.focus();
    const content = cellRef.current!.textContent;

    //changes cursor position
    setCursorPositionInACell(content, cellRef, contentPaddingLeft, e)
  }

  function disableEditing() {
    cellRef.current!.contentEditable = 'false';
    setCursor('cursorCell');
    changeSelected('ordinary', address);
  }

  function selectCell() {
    changeSelected('selected', address);
  }


  return (
    <div
      id={address}
      style={cellStyle}
      data-testid={address}
      onDoubleClick={(e) => enableEditing(e)}
      onClick={() => selectCell()}
      onBlur={disableEditing}
      className={`cell ${cursor}`}
    >
      <div
        ref={cellRef}
        data-testid="content"
        spellCheck="false"
        className="content"
        style={contentStyles}
      >
        {content}
      </div>
    </div>
  );
};

export const Cell = memo(InitialCell);
