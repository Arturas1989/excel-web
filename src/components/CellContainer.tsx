import { useWorkbookContext } from '../context/useWorkbookContext';
import { Cells } from '../components';
import { useMethods } from '../hooks/useMethods';

export const CellContainer = () => {
  const workbookContext = useWorkbookContext();
  const { selectionGrid, cellsObj } = workbookContext;
  const { rows, cols } = cellsObj;

  const methods = useMethods(workbookContext);

  const { grid } = selectionGrid;

  return (
    <div className="cell-container">
      <Cells grid={grid} rows={rows} cols={cols} methods={methods}/>
    </div>
  );
};
