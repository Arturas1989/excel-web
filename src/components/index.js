export { Cell } from './Cell';
export { CellRow } from './CellRow';
export { Cells } from './Cells';
export { CellContainer } from './CellContainer';