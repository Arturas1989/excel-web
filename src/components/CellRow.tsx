import { memo } from "react";
import { CellType, Methods } from "../types";
import { Cell } from "./Cell";
import {isEqual} from 'lodash';

type CellRowProps = {
  row: CellType[];
  methods: Methods;
  index: number;
};

const InitialCellRow = ({ row, methods, index }: CellRowProps) => {
  console.log('rendered row: ' + index)
  return (
    <div className="cell-row">
      {row.map(cell => {
        return (
          <Cell
            key={cell.address}
            cell={cell}
            methods={methods}
            cellStyle={cell.cellStyles}
          />
        );
      })}
    </div>
  );
};

const propsAreEqual = (oldProps: CellRowProps, newProps: CellRowProps) => {
  let prop: keyof CellRowProps;
  for(prop in oldProps){
    if(!isEqual(oldProps[prop], newProps[prop])) return false;
  }
  return true;
}

export const CellRow = memo(InitialCellRow, propsAreEqual);
