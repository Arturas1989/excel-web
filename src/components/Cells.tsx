import { Methods, Grid } from '../types';
import { getAddress } from '../utils/helpers/cellHelper';
import { CellRow } from './CellRow';

type CellRowProps = {
  grid: Grid;
  rows: number;
  cols: number;
  methods: Methods;
};

export const Cells = ({ grid, rows, cols, methods }: CellRowProps) => {
  let cells = [];
  for (let i = 0; i < rows; i++) {
    let row = [];
    for (let j = 0; j < cols; j++) {
      const address = getAddress(i, j);
      if(grid[address]) row.push(grid[address]);
    }
    cells.push(row);
  }

  return (
    <>
      {cells.map((row, i) => {
        return <CellRow key={i} row={row} methods={methods} index={i}/>
      })}
    </>
  );
};
