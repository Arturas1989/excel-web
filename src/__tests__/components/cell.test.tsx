import renderer from 'react-test-renderer';
import { fireEvent, render, screen } from '@testing-library/react';
import { Cell } from '../../components';
import { Cells } from '../../Workbook/Cells'

const getCell = () => {
  const cells = new Cells();
  const grid = cells.grid;
  return grid[0][0];
}

describe('Cell on click and render', () => {
  const gridCell = getCell();
  test('Cell snapshot', () => {
    
    const tree = renderer
      .create(
        <Cell cell={gridCell} />
      )
      .toJSON();
    expect(tree).toMatchSnapshot();
  });

  test('Cell is in the document', () => {
    render(<Cell cell={gridCell} />)
    const cell = screen.getAllByTestId('cell')[0];
    expect(cell).toBeInTheDocument();
  });

  test('Cell on click has a selected class', () => {
    render(<Cell cell={gridCell} />)
    const cell = screen.getAllByTestId('cell')[0];
    expect(cell).not.toHaveClass('selected');

    fireEvent.click(cell);
    expect(cell).toHaveClass('selected');
  });
});

describe('Cell on double click', () => {
  const gridCell = getCell();
  test('Cell has a cursorText class', () => {
    render(<Cell cell={gridCell} />)
    const cell = screen.getAllByTestId('cell')[0];
    expect(cell).not.toHaveClass('cursorText');

    fireEvent.doubleClick(cell);
    expect(cell).toHaveClass('cursorText');
  });

  test('Cell has a selected class', () => {
    render(<Cell cell={gridCell} />)
    const cell = screen.getAllByTestId('cell')[0];
    expect(cell).not.toHaveClass('selected');

    fireEvent.doubleClick(cell);
    expect(cell).toHaveClass('selected');
  });
});

describe('Cell on blur', () => {
  const gridCell = getCell();
  test('Cell has a cursorCell class', () => {
    render(<Cell cell={gridCell} />)
    const cell = screen.getAllByTestId('cell')[0];
    fireEvent.blur(cell);
    expect(cell).toHaveClass('cursorCell');
  });

  test('Cell has no selected class', () => {
    render(<Cell cell={gridCell} />)
    const cell = screen.getAllByTestId('cell')[0];
    fireEvent.blur(cell);
    expect(cell).not.toHaveClass('selected');
  });
});