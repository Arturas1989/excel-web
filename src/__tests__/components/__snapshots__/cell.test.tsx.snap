// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Cell on click and render Cell snapshot 1`] = `
<div
  className="cell cursorCell "
  data-testid="cell"
  id="A1"
  onBlur={[Function]}
  onClick={[Function]}
  onDoubleClick={[Function]}
  style={
    Object {
      "border": "1px solid rgb(212, 212, 212)",
      "height": "34px",
      "overflow": "clip",
      "padding": "8px",
      "width": "100px",
    }
  }
>
  <div
    className="content"
    data-testid="content"
    spellCheck="false"
    style={
      Object {
        "fontSize": "16px",
        "paddingLeft": "1px",
      }
    }
  />
</div>
`;
